### ISSUE TITLE
[BUG] Backup fails to complete due to timeout causing duplicate backups

### ISSUE THREAD

**Issue by tyzbit** [longhorn/longhorn#2278]
**Describe the bug**
Scheduled backup to S3 results in multiple copies of each individual backup in S3.  Furthermore, automatic cleanup appears to be broken.

**To Reproduce**

Prerequisite: Configure S3 Backups
1. Create a volume in Longhorn
2. Configure snapshots twice a day, retaining 1 (not necessary).
3. Configure backups twice a day, retaining 4.

**Expected behavior**
1 snapshot is retained (this works)
Each backup results in one backup
4 backups are retained and any older backups are cleaned up.

**Actual behavior**
1 snapshot is retained
Each backup results in 4 backups
Old backups are not cleaned up

**Log & Screenshot**
<img width="614" alt="image" src="https://user-images.githubusercontent.com/3319104/109404462-536f5e80-7934-11eb-8a9f-6daba907dcf7.png">
<img width="1389" alt="image" src="https://user-images.githubusercontent.com/3319104/109404468-684bf200-7934-11eb-9ab2-0a4aadc3749d.png">
(I cleaned up many of the older backups)

[Support Bundle](https://drive.google.com/file/d/1yMIqi_ZBoE4UWxMcvGfJG2-x1ILP8R00/view)

**Environment:**
 - Longhorn version: v1.1.0
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: kubeadm, v1.19.7
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 6 total, 3 with replicas
 - Node config
   - OS type and version: Ubuntu 20.04.1 LTS
   - CPU per node: 8
   - Memory per node: 16GB
   - Disk type(e.g. SSD/NVMe): NVMe (Samsung 970 Evo Plus)
   - Network bandwidth between the nodes: 1Gbps
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal
 - Number of Longhorn volumes in the cluster: 15

**Comment by tabnul** [comment]
Do your backups even complete effectively? 
I also have this issue since a week but i think it is due to some backups suddenly starting crashing (cleanup not being kicked off afterwards, or cleanup causes crashing?)

```
2021-03-01 02:02:00	
time="2021-03-01T01:02:00Z" level=info msg="Running recurring backup for volume pvc-1129530a-ceb8-4c33-9072-47144695729a"
2021-03-01 02:01:24	
time="2021-03-01T01:01:24Z" level=fatal msg="Error taking snapshot: failed to complete backupAndCleanup for pvc-1129530a-ceb8-4c33-9072-47144695729a: Post \"http://longhorn-backend:9500/v1/volumes/pvc-1129530a-ceb8-4c33-9072-47144695729a?action=snapshotCreate\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
2021-03-01 02:01:24	
time="2021-03-01T01:01:24Z" level=info msg="Handling volume detachment for volume pvc-1129530a-ceb8-4c33-9072-47144695729a"
2021-03-01 02:01:14	
time="2021-03-01T01:01:14Z" level=info msg="Running recurring backup for volume pvc-1129530a-ceb8-4c33-9072-47144695729a"
2021-03-01 02:00:57	
time="2021-03-01T01:00:57Z" level=fatal msg="Error taking snapshot: failed to complete backupAndCleanup for pvc-1129530a-ceb8-4c33-9072-47144695729a: Post \"http://longhorn-backend:9500/v1/volumes/pvc-1129530a-ceb8-4c33-9072-47144695729a?action=snapshotCreate\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
2021-03-01 02:00:57	
time="2021-03-01T01:00:57Z" level=info msg="Handling volume detachment for volume pvc-1129530a-ceb8-4c33-9072-47144695729a"
2021-03-01 02:00:47	
time="2021-03-01T01:00:47Z" level=info msg="Running recurring backup for volume pvc-1129530a-ceb8-4c33-9072-47144695729a"
2021-03-01 02:00:36	
time="2021-03-01T01:00:36Z" level=fatal msg="Error taking snapshot: failed to complete backupAndCleanup for pvc-1129530a-ceb8-4c33-9072-47144695729a: Post \"http://longhorn-backend:9500/v1/volumes/pvc-1129530a-ceb8-4c33-9072-47144695729a?action=snapshotCreate\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
2021-03-01 02:00:36	
time="2021-03-01T01:00:36Z" level=info msg="Handling volume detachment for volume pvc-1129530a-ceb8-4c33-9072-47144695729a"
2021-03-01 02:00:26	
time="2021-03-01T01:00:26Z" level=info msg="Running recurring backup for volume pvc-1129530a-ceb8-4c33-9072-47144695729a"

```

**Comment by joshimoo** [comment]
I am guessing the timeout will mark the job as failed.
Even though the backup is actually running and will complete successfully.
Thanks for reporting we will look into this.

**Comment by tabnul** [comment]
> I am guessing the timeout will mark the job as failed.
> Even though the backup is actually running and will complete successfully.
> Thanks for reporting we will look into this.

keep in mind that the logfile i posted might not be related to the problem of @tyzbit .
it does however seems to be a 'possible' scenario.
@tyzbit do you see failed backjobs ? (kubectl get job )

**Comment by tyzbit** [comment]
I see logs identical to the ones you posted, so it appears we have the same issue.

**Comment by shuo-wu** [comment]
I cannot find the root cause. The related backup pods as well as some old replica manager pods are gone, and the info in the longhorn manager is insufficient. This bug maybe not related to the snapshot/backup timeout. There is no timeout error for all snapshot creation and backup creation. And the interval between 2 snapshot creations or backup creations is less than 1 minute, which is the timeout that triggers https://github.com/longhorn/longhorn/issues/2029. I guess the recurring backup pods may get crashed quickly due to some unknown reasons. 


I do find some weird logs in the support bundle but I don't think it's the cause:
1. There are some snapshot deletion failure logs during the last 4 backup creations. These errors should block the backup creation job rather than duplicate backups. But in fact, the result is the opposite...
```
2021-02-27T17:50:04.929953182Z time="2021-02-27T17:50:04Z" level=debug msg="Created snapshot c-xy9ruf-2c24d392 with labels map[RecurringJob:c-xy9ruf] for volume pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e"
2021-02-27T17:50:04.929966981Z 192.168.133.82 - - [27/Feb/2021:17:50:04 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotCreate HTTP/1.1" 200 365 "" "Go-http-client/1.1"
2021-02-27T17:50:04.959459705Z 192.168.133.82 - - [27/Feb/2021:17:50:04 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotList HTTP/1.1" 200 3336 "" "Go-http-client/1.1"
2021-02-27T17:50:04.978328309Z time="2021-02-27T17:50:04Z" level=warning msg="HTTP handling error fail to delete snapshot: error deleting snapshot 'c-xy9ruf-15dab7c9': Failed to execute: /var/lib/longhorn/engine-binaries/rancher-longhornio-longhorn-engine-v1.1.0/longhorn [--url 192.168.187.205:10003 snapshot rm c-xy9ruf-15dab7c9], output , stderr, Failed to delete c-xy9ruf-15dab7c9: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\ntime=\"2021-02-27T17:50:04Z\" level=fatal msg=\"Error running rm snapshot command: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\"\n, error exit status 1"
2021-02-27T17:50:04.978375432Z time="2021-02-27T17:50:04Z" level=error msg="Error in request: fail to delete snapshot: error deleting snapshot 'c-xy9ruf-15dab7c9': Failed to execute: /var/lib/longhorn/engine-binaries/rancher-longhornio-longhorn-engine-v1.1.0/longhorn [--url 192.168.187.205:10003 snapshot rm c-xy9ruf-15dab7c9], output , stderr, Failed to delete c-xy9ruf-15dab7c9: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\ntime=\"2021-02-27T17:50:04Z\" level=fatal msg=\"Error running rm snapshot command: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\"\n, error exit status 1"
......
2021-02-27T17:50:17.779848128Z time="2021-02-27T17:50:17Z" level=debug msg="Created snapshot c-xy9ruf-22328083 with labels map[RecurringJob:c-xy9ruf] for volume pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e"
2021-02-27T17:50:17.779948033Z 192.168.133.82 - - [27/Feb/2021:17:50:17 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotCreate HTTP/1.1" 200 340 "" "Go-http-client/1.1"
2021-02-27T17:50:17.802441438Z 192.168.133.82 - - [27/Feb/2021:17:50:17 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotList HTTP/1.1" 200 3682 "" "Go-http-client/1.1"
2021-02-27T17:50:17.853705934Z 192.168.133.82 - - [27/Feb/2021:17:50:17 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotDelete HTTP/1.1" 500 737 "" "Go-http-client/1.1"
2021-02-27T17:50:17.853708852Z time="2021-02-27T17:50:17Z" level=warning msg="HTTP handling error fail to delete snapshot: error deleting snapshot 'c-xy9ruf-15dab7c9': Failed to execute: /var/lib/longhorn/engine-binaries/rancher-longhornio-longhorn-engine-v1.1.0/longhorn [--url 192.168.187.205:10003 snapshot rm c-xy9ruf-15dab7c9], output , stderr, Failed to delete c-xy9ruf-15dab7c9: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\ntime=\"2021-02-27T17:50:17Z\" level=fatal msg=\"Error running rm snapshot command: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\"\n, error exit status 1"
2021-02-27T17:50:17.853781178Z time="2021-02-27T17:50:17Z" level=error msg="Error in request: fail to delete snapshot: error deleting snapshot 'c-xy9ruf-fdc84027': Failed to execute: /var/lib/longhorn/engine-binaries/rancher-longhornio-longhorn-engine-v1.1.0/longhorn [--url 192.168.187.205:10003 snapshot rm c-xy9ruf-15dab7c9], output , stderr, Failed to delete c-xy9ruf-15dab7c9: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\ntime=\"2021-02-27T17:50:17Z\" level=fatal msg=\"Error running rm snapshot command: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\"\n, error exit status 1"
......
2021-02-27T17:50:43.075437697Z time="2021-02-27T17:50:43Z" level=debug msg="Created snapshot c-xy9ruf-cbacd94a with labels map[RecurringJob:c-xy9ruf] for volume pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e"
2021-02-27T17:50:43.075587752Z 192.168.133.82 - - [27/Feb/2021:17:50:42 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotCreate HTTP/1.1" 200 340 "" "Go-http-client/1.1"
2021-02-27T17:50:43.104589377Z 192.168.133.82 - - [27/Feb/2021:17:50:43 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotList HTTP/1.1" 200 4028 "" "Go-http-client/1.1"
2021-02-27T17:50:43.157686400Z time="2021-02-27T17:50:43Z" level=warning msg="HTTP handling error fail to delete snapshot: error deleting snapshot 'c-xy9ruf-15dab7c9': Failed to execute: /var/lib/longhorn/engine-binaries/rancher-longhornio-longhorn-engine-v1.1.0/longhorn [--url 192.168.187.205:10003 snapshot rm c-xy9ruf-15dab7c9], output , stderr, Failed to delete c-xy9ruf-15dab7c9: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\ntime=\"2021-02-27T17:50:43Z\" level=fatal msg=\"Error running rm snapshot command: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\"\n, error exit status 1"
2021-02-27T17:50:43.157771781Z time="2021-02-27T17:50:43Z" level=error msg="Error in request: fail to delete snapshot: error deleting snapshot 'c-xy9ruf-15dab7c9': Failed to execute: /var/lib/longhorn/engine-binaries/rancher-longhornio-longhorn-engine-v1.1.0/longhorn [--url 192.168.187.205:10003 snapshot rm c-xy9ruf-15dab7c9], output , stderr, Failed to delete c-xy9ruf-15dab7c9: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\ntime=\"2021-02-27T17:50:43Z\" level=fatal msg=\"Error running rm snapshot command: Can not remove a snapshot because tcp://192.168.133.123:10195 is rebuilding\"\n, error exit status 1"
2021-02-27T17:50:43.157704674Z 192.168.133.82 - - [27/Feb/2021:17:50:43 +0000] "POST /v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotDelete HTTP/1.1" 500 737 "" "Go-http-client/1.1"
2021-02-27T17:50:43.662840296Z time="2021-02-27T17:50:43Z" level=debug msg="Skip rebuilding of replica because there is another rebuild in progress" controller=longhorn-engine node=showman volume=pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e
2021-02-27T17:50:50.426341217Z time="2021-02-27T17:50:50Z" level=debug msg="Backup backup-430369d81c954123 created for volume pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e snapshot c-xy9ruf-2c24d392"
2021-02-27T17:50:50.426435634Z time="2021-02-27T17:50:50Z" level=debug msg="Initiated Backup backup-430369d81c954123 for snapshot c-xy9ruf-2c24d392 of volume pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e with label map[KubernetesStatus:{\"pvName\":\"pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e\",\"pvStatus\":\"Bound\",\"namespace\":\"default\",\"pvcName\":\"nextcloud-data\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"nextcloud-c4fd65ff4-phfdp\",\"podStatus\":\"Running\",\"workloadName\":\"nextcloud-c4fd65ff4\",\"workloadType\":\"ReplicaSet\"}],\"lastPodRefAt\":\"\"} RecurringJob:c-xy9ruf]"
......
2021-02-27T17:51:04.131535698Z time="2021-02-27T17:51:04Z" level=debug msg="Backup backup-94c901cf08ab4eac created for volume pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e snapshot c-xy9ruf-22328083"
2021-02-27T17:51:04.131633552Z time="2021-02-27T17:51:04Z" level=debug msg="Initiated Backup backup-94c901cf08ab4eac for snapshot c-xy9ruf-22328083 of volume pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e with label map[KubernetesStatus:{\"pvName\":\"pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e\",\"pvStatus\":\"Bound\",\"namespace\":\"default\",\"pvcName\":\"nextcloud-data\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"nextcloud-c4fd65ff4-phfdp\",\"podStatus\":\"Running\",\"workloadName\":\"nextcloud-c4fd65ff4\",\"workloadType\":\"ReplicaSet\"}],\"lastPodRefAt\":\"\"} RecurringJob:c-xy9ruf]"
```

For Longhorn side, I think we need to broadcast the events for the recurring job pods. So that we can check if the pod succeeded, or failed due to some reason.

**Comment by muhlba91** [comment]
similar problem here but i receive a very odd behaviour.

number of volumes: 3
cluster created via kubeadm, running on raspberry pi's (arm64)
backup policy: twice a day, retain 4
backup to S3: backblaze

volume 1 created 8 days ago (size 5Gi, actual size ~750MB) has each backup showing up multiple times. in contrast to @tyzbit i don't see a pattern here. sometimes it's showing up as 2 backups, sometimes as 3 and sometimes as 4. i didn't observe more than that. overall i have 41 backups since 8 days. without retention settings, this should max out at 16, with the retain option set it should no be more than 4.

volume 1 created 5 days ago (size 128Mi, actual size ~44MB) has a completely different behaviour to volume 1 with retention working (i only have 4 backups showing) but each backup is duplicated. that said, i have 2 backups of time X and 2 backups of time Y, totalling to 4 backups as per the retain option.

volume 3 created 43 mins ago (just before the backup schedule started, size 128Mi, actual size ~26MB) and should only have 1 backup existing at all. however, it shows this backup 3 times, so i see 3 backups at the same time.

i also saw 2 jobs getting killed by kubernetes due to backoff (`Job has reached the specified backoff limit`) which belong to exactly those volumes where the backups show up multiple times.
from this observation it looks like backups are done successfully but afterwards, when the cleanup happens, the jobs fail (no logs left due to kubernetes terminating the pod). looking at the volume sizes i'm not sure if a timeout can happen as the last volume is very small and shows the same problem.

**Comment by shuo-wu** [comment]
> from this observation it looks like backups are done successfully but afterwards, when the cleanup happens, the jobs fail (no logs left due to kubernetes terminating the pod). looking at the volume sizes i'm not sure if a timeout can happen as the last volume is very small and shows the same problem.

@muhlba91 Thanks for your information! I guess [this backup listing timeout bug](https://github.com/longhorn/longhorn/issues/1955) before the cleanup may fail the job. If there are multiple backup volumes in your backupstore, this may be the cause.

**Comment by tyzbit** [comment]
Just as an update, I pruned my volumes down to 10 volumes total and scheduled a backup on the problematic volume which completed successfully with only one backup saved.  I am also going to upgrade to 2.5GbE tomorrow which should put me in spec for the recommended network throughput for 10 volumes.

**Comment by PhanLe1010** [comment]
Regarding the question https://github.com/longhorn/longhorn/issues/2278#issuecomment-791173613, when we failed to delete a snapshot, we return `nil` and continue to take the backup here:
https://github.com/longhorn/longhorn-manager/blob/f6ff18666ea238c4eec4bfdce9612796da1a0701/app/volume.go#L297
I think this is intended behavior , not a bug  because we would rather have extra backup than missing backup.

For the reason why deleting snapshot fail, it is because of volume is rebuilding. For why the volume is rebuilding, it is likely because of the network not being fast enough. This is confirmed by https://github.com/longhorn/longhorn/issues/2278#issuecomment-804913204

**Comment by tyzbit** [comment]
I believe I spoke too soon.  When I did a manual backup it completed successfully however I re-enabled scheduled backups and the issue is back with 3 backups for every backup.
<img width="1384" alt="image" src="https://user-images.githubusercontent.com/3319104/112566838-19b83900-8db6-11eb-8336-2006f2c65147.png">

I upgraded to 2.5GbE between the nodes as well, which puts me in spec for recommended bandwidth with respect to the number of volumes I have.

One thing I don't think has been brought up is guaranteed engine CPU - I discovered I'm currently at 0.5 for 10 volumes (I thought I had increased this).  I will increase this to 1.0 and will report back my results.

**Comment by tyzbit** [comment]
Increasing the guaranteed CPU to 1.0 had little effect; my cluster has a good amount of spare CPU per node so Longhorn would have had more CPU when it needed it.

I actually happened to catch the backup run about 5 minutes after it started and it looks like two backups are started basically at once
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/3319104/112785132-4cb23500-9021-11eb-987c-0a5267ef42f0.png">

Backup configuration
<img width="621" alt="image" src="https://user-images.githubusercontent.com/3319104/112785153-59368d80-9021-11eb-858e-f75804329a63.png">

[Here are the logs](https://gist.github.com/tyzbit/9a41c9463fcad136397738bbc017c6f7) from the 2 minutes when those backups started.  I knew the backup failed and thats why a second one was created, but I didn't realize those events happened in less than a minute.  That doesn't seem like very long to wait for the backup to finish.

**Comment by shuo-wu** [comment]
```
2021-03-29 03:50:50 | pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e-c-xy9ruf-c | cameraman | pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e-c-xy9ruf-c-161698r5nnc | time="2021-03-29T03:50:50Z" level=fatal msg="Error taking snapshot: failed to complete backupAndCleanup for pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e: Post "http://longhorn-backend:9500/v1/volumes/pvc-77903a7e-7c93-4522-8f72-6e9c5c8dff7e?action=snapshotBackup\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"

```

Found some backup failure caused by the timeout. Need to check how long the timeout is and how the timeout can be triggered in v1.1.0 & v1.1.1.

**Comment by innobead** [comment]
This fix needs backport to 1.1.2 as well, and it is already in the commit history.

ref: https://github.com/longhorn/longhorn-manager/commits/v1.1.2

**Comment by tyzbit** [comment]
Just want to close the loop on this and confirm that by updating to `1.1.2` backups are working as expected.  Good work!
